src/lib.rs

```diff
@@ -531,14 +658,61 @@ mod tests {
     }
 
     #[test]
+    fn memchr_iter() {
+        let haystack = b"aaaabaaaab";
+        let mut memchr_iter = Memchr::new(b'b', haystack);
+        let first = memchr_iter.next();
+        let second = memchr_iter.next();
+        let third = memchr_iter.next();
+        assert_eq!(Some(5), first);
+        assert_eq!(Some(10), second);
+        assert_eq!(None, third);
+    }
+
+    #[test]
+    fn memchr2_iter() {
+        let haystack = b"ab";
+        let mut memchr_iter = Memchr2::new(b'a', b'b', haystack);
+        let first = memchr_iter.next();
+        let second = memchr_iter.next();
+        let third = memchr_iter.next();
+        assert_eq!(Some(1), first);
+        assert_eq!(Some(2), second);
+        assert_eq!(None, third);
+    }
+
+    #[test]
+    fn memchr3_iter() {
+        let haystack = b"abc";
+        let mut memchr_iter = Memchr3::new(b'a', b'b', b'c', haystack);
+        let first = memchr_iter.next();
+        let second = memchr_iter.next();
+        let third = memchr_iter.next();
+        let fourth = memchr_iter.next();
+        assert_eq!(Some(1), first);
+        assert_eq!(Some(2), second);
+        assert_eq!(Some(3), third);
+        assert_eq!(None, fourth);
+    }
+
+    #[test]
+    fn memchr_reverse_iter() {
+        let haystack = b"aaaabaaaab";
+        let mut memchr_iter = Memchr::new(b'b', haystack);
+        let first = memchr_iter.next();
+        let second = memchr_iter.next_back();
+        let third = memchr_iter.next();
+
+        assert_eq!(Some(5), first);
+        assert_eq!(Some(10), second);
+        assert_eq!(None, third);
```

Comment:
Yeah I think that 3rd `b` did the function in.  I don't think the output is correct.
```
let haystack = b"aaaabaaaabaaaab";
let mut memchr_iter = Iter::new(b'b', haystack);
let first = memchr_iter.next();
println!("first: {:?}", first);
let second = memchr_iter.next_back();
println!("second: {:?}", second);
let third = memchr_iter.next();
println!("third: {:?}", third);
let fourth = memchr_iter.next_back();
println!("fourth: {:?}", fourth);
//output
first: Some(5)
second: Some(15)
third: Some(20)
fourth: None
```